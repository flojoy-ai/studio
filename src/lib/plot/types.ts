export interface Drawable {
  draw: () => void;
}
