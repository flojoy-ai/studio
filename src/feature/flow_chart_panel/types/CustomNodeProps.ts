export type ElementsData = {
  id: string;
  label: string;
  func: string;
  type: string;
  running?: boolean;
  failed?:boolean;
  ctrls: {
    [key: string]: {
      functionName: string;
      param: string;
      value: number;
    };
  };
  inputs?: Array<{ name: string; id: string; type:string }>
  selects?: any;
}

export interface CustomNodeProps {
  data: ElementsData;
}