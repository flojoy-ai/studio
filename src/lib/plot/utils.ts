export function clamp(x: number, lo: number, hi: number) {
  return Math.min(Math.max(x, lo), hi);
}
