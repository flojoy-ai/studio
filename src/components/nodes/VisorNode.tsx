import { CustomNodeProps } from "@src/types/node";
import NodeWrapper from "@/components/common/NodeWrapper";
import clsx from "clsx";
import HandleComponent from "@/components/common/HandleComponent";
import Scatter from "@/assets/nodes/Scatter";
import { memo, useMemo } from "react";
import { makePlotlyData } from "@/components/plotly/formatPlotlyData";
import PlotlyComponent from "@/components/plotly/PlotlyComponent";
import CompositePlot from "@/assets/nodes/CompositePlot";
import ProphetComponents from "@/assets/nodes/ProphetComponents";
import ProphetPlot from "@/assets/nodes/ProphetPlot";
import ArrayView from "@/assets/nodes/ArrayView";
import MatrixView from "@/assets/nodes/MatrixView";
import BigNumber from "@/assets/nodes/BigNumber";
import BoxPlot from "@/assets/nodes/BoxPlot";
import Histogram from "@/assets/nodes/Histogram";
import LineChart from "@/assets/nodes/LineChart";
import Scatter3D from "@/assets/nodes/3DScatter";
import Surface3D from "@/assets/nodes/3DSurface";
import Bar from "@/assets/nodes/Bar";
import Table from "@/assets/nodes/Table";
import Image from "@/assets/nodes/Image";
import MarkDownText from "@/components/common/MarkDownText";
import PeakFinder from "@/assets/nodes/PeakFinder";
import RegionInspector from "@/assets/nodes/RegionInspector";
import { useTheme } from "@src/providers/theme-provider";

const chartElemMap: { [func: string]: React.JSX.Element } = {
  SCATTER: <Scatter />,
  HISTOGRAM: <Histogram />,
  LINE: <LineChart />,
  SURFACE3D: <Surface3D />,
  SCATTER3D: <Scatter3D />,
  BAR: <Bar />,
  TABLE: <Table />,
  IMAGE: <Image />,
  BOX: <BoxPlot />,
  BIG_NUMBER: <BigNumber />,
  MATRIX_VIEW: <MatrixView />,
  ARRAY_VIEW: <ArrayView />,
  PROPHET_PLOT: <ProphetPlot />,
  PROPHET_COMPONENTS: <ProphetComponents />,
  COMPOSITE: <CompositePlot />,
  TEXT_VIEW: <Table />,
  EXTREMA_DETERMINATION: <PeakFinder />,
  REGION_PROPERTIES: <RegionInspector />,
};

const VisorNode = (props: CustomNodeProps) => {
  const {
    nodeProps: { data },
    nodeError,
    isRunning,
    plotlyFig,
    textBlob,
  } = props;

  const { resolvedTheme } = useTheme();

  const plotlyData = useMemo(
    () =>
      plotlyFig ? makePlotlyData(plotlyFig.data, resolvedTheme, true) : null,
    [plotlyFig, resolvedTheme],
  );

  return (
    <NodeWrapper wrapperProps={props}>
      <div
        className={clsx(
          "rounded-2xl bg-transparent",
          { "shadow-around shadow-accent2": isRunning || data.selected },
          { "shadow-around shadow-red-700": nodeError },
        )}
      >
        {plotlyData && (
          <PlotlyComponent
            data={plotlyData}
            id={data.id}
            layout={plotlyFig?.layout ?? {}}
            useResizeHandler
            style={{
              height: 293,
              width: 380,
            }}
            isThumbnail
          />
        )}
        {textBlob && <MarkDownText text={textBlob} isThumbnail />}
        {!plotlyData && !textBlob && <>{chartElemMap[data.func]}</>}
        <HandleComponent data={data} variant="accent2" />
      </div>
    </NodeWrapper>
  );
};

export default memo(VisorNode);
