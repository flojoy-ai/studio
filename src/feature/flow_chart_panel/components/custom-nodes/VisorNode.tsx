import HandleComponent from "@feature/flow_chart_panel/components/HandleComponent";
import { CustomNodeProps } from "@feature/flow_chart_panel/types/CustomNodeProps";
import { useFlowChartState } from "@hooks/useFlowChartState";
import { Box, clsx, createStyles, useMantineTheme } from "@mantine/core";
import PlotlyComponent from "@src/feature/common/PlotlyComponent";
import { useSocket } from "@src/hooks/useSocket";
import { Layout } from "plotly.js";
import { useEffect } from "react";
import { BGTemplate } from "../../svgs/histo-scatter-svg";
import { useNodeStyles } from "../DefaultNode";
import NodeWrapper from "../NodeWrapper";
import Scatter3D from "../nodes/3d-scatter";
import Surface3D from "../nodes/3d-surface";
import Histogram from "../nodes/Histogram";
import Scatter from "../nodes/Scatter";
import BarChart from "../nodes/bar";
import LineChart from "../nodes/line-chart";
import usePlotLayout from "@src/feature/common/usePlotLayout";
import { useMemo } from "react";

const useStyles = createStyles((theme) => {
  return {
    visorNode: {
      background: "transparent",
      color:
        theme.colorScheme === "light"
          ? theme.colors.accent1[0]
          : theme.colors.accent2[0],
    },
  };
});

const chartElemMap: { [func: string]: JSX.Element } = {
  SCATTER: <Scatter />,
  HISTOGRAM: <Histogram />,
  LINE: <LineChart />,
  SURFACE3D: <Surface3D />,
  SCATTER3D: <Scatter3D />,
  BAR: <BarChart />,
};

const VisorNode = ({ data }: CustomNodeProps) => {
  const nodeClasses = useNodeStyles().classes;
  const { classes } = useStyles();
  const theme = useMantineTheme();
  const { runningNode, failedNode, selectedNode } = useFlowChartState();
  const params = data.inputs || [];

  const selected = selectedNode ? selectedNode.id === data.id : false;

  // TODO: Investigate why this keeps making it rerender
  const { states } = useSocket();
  const { programResults } = states!;

  const results = programResults?.io;
  const result = results?.find((r) => r.id === data.id);

  const plotLayout = usePlotLayout();

  const accentColor =
    theme.colorScheme === "dark"
      ? theme.colors.accent1[0]
      : theme.colors.accent2[0];

  const layoutOverride: Partial<Layout> = {
    plot_bgcolor: "transparent",
    title: data.label,
    margin: { t: 30, r: 0, b: 0, l: 0 },
    grid: { rows: 0, columns: 0 },
    xaxis: { visible: false },
    yaxis: { visible: false },
    font: {
      color: accentColor,
    },
  };

  const plotlyResultData = useMemo(
    () =>
      result
        ? result.result.default_fig.data.map((d) => ({
            ...d,
            marker: {
              ...d.marker,
              color: accentColor,
            },
          }))
        : undefined,
    [result]
  );

  return (
    <NodeWrapper data={data}>
      <Box
        className={clsx(
          runningNode === data.id || selected ? nodeClasses.defaultShadow : "",
          failedNode === data.id ? nodeClasses.failShadow : ""
        )}
      >
        {result && plotlyResultData ? (
          <>
            <PlotlyComponent
              data={plotlyResultData}
              id={data.id}
              layout={{ ...plotLayout, ...layoutOverride }}
              useResizeHandler
              style={{
                height: 190,
                width: 210,
              }}
            />

            <Box
              display="flex"
              h={params.length > 0 ? (params.length + 1) * 40 : "fit-content"}
              sx={{
                flexDirection: "column",
              }}
            >
              <HandleComponent data={data} inputs={params} />
            </Box>
          </>
        ) : (
          <Box
            className={clsx(classes.visorNode, nodeClasses.nodeContainer)}
            style={{
              ...(params.length > 0 && { padding: "0px 0px 8px 0px" }),
            }}
          >
            {chartElemMap[data.func]}
            <BGTemplate theme={theme.colorScheme} />
            <Box
              display="flex"
              h={params.length > 0 ? (params.length + 1) * 40 : "fit-content"}
              sx={{
                flexDirection: "column",
              }}
            >
              <HandleComponent data={data} inputs={params} />
            </Box>
          </Box>
        )}
      </Box>
    </NodeWrapper>
  );
};

export default VisorNode;
