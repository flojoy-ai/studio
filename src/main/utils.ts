import * as http from "http";
import * as fs from "fs";
import { isIP } from "net";
import { execCommand } from "./executor";
import { Command } from "./command";
import { app, dialog } from "electron";
import { join } from "path";
import { killCaptain } from "./python";
import log from "electron-log";
import { ChildProcess } from "node:child_process";
import { Err, Result, Ok } from "../types/result";

export const isPortFree = (port: number) =>
  new Promise((resolve) => {
    const server = http
      .createServer()
      .listen(port, "127.0.0.1", () => {
        server.close();
        resolve(true);
      })
      .on("error", () => {
        resolve(false);
      });
  });

export const killProcess = async (port: number) => {
  await execCommand(
    new Command({
      darwin: `kill -9 $(lsof -t -i :${port})`,
      linux: `kill -9 $(lsof -t -i :${port})`,
      win32: `FOR /F "tokens=5" %i IN ('netstat -aon ^| find "${port}"') DO Taskkill /F /PID %i`,
    }),
  );
};

export const ping = async (addr: string) => {
  if (isIP(addr) === 0) {
    throw new Error(`Invalid IP address: ${addr}`);
  }

  return await execCommand(
    new Command({
      win32: `ping -n 1 ${addr}`,
      darwin: `ping -c 1 ${addr}`,
      linux: `ping -c 1 ${addr}`,
    }),
    { quiet: true },
  );
};

export const netstat = async () => {
  return await execCommand(
    new Command({
      win32: "netstat",
      darwin: "netstat",
      linux: "netstat",
    }),
    { quiet: true },
  );
};

export const ifconfig = async () => {
  return await execCommand(
    new Command({
      win32: "ipconfig /all",
      darwin: "ifconfig",
      linux: "ifconfig",
    }),
    { quiet: true },
  );
};

export const writeFileSync = (_, filePath: string, text: string): void => {
  fs.writeFileSync(filePath, text);
};

export const pickDirectory = async (): Promise<string> => {
  const handler = await dialog.showOpenDialog({
    properties: ["openDirectory"],
  });
  return handler.canceled ? "" : handler.filePaths[0];
};

export const getCustomBlocksDir = async () => {
  const filePath = join(app.getPath("home"), ".flojoy/custom_blocks_path.txt");

  if (!fs.existsSync(filePath)) {
    return undefined;
  }

  const blocksPath = fs
    .readFileSync(filePath, { encoding: "utf-8" })
    .toString();

  if (!fs.existsSync(blocksPath)) {
    return undefined;
  }

  return blocksPath;
};

export const cacheCustomBlocksDir = (_, dirPath: string) => {
  const flojoyDir = join(app.getPath("home"), ".flojoy");
  if (!fs.existsSync(flojoyDir)) {
    fs.mkdirSync(flojoyDir);
  }
  const cacheFilePath = join(flojoyDir, "custom_blocks_path.txt");
  fs.writeFileSync(cacheFilePath, dirPath);
};

export const openTestPicker = (): Promise<
  { filePath: string; fileContent: string } | undefined
> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: ["json", "py"],
            name: "Test",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const openFilePicker = (
  _,
  allowedExtensions: string[] = ["json"],
): Promise<{ filePath: string; fileContent: string } | undefined> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: allowedExtensions,
            name: "allowed extensions",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
        resolve(undefined);
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const cleanup = async () => {
  const captainProcess = global.captainProcess as ChildProcess;
  log.info(
    "Cleanup function invoked, is captain running? ",
    !(captainProcess?.killed ?? true),
  );
  if (captainProcess && captainProcess.exitCode === null) {
    const success = killCaptain();
    if (success) {
      global.captainProcess = null;
      log.info("Successfully terminated captain :)");
    } else {
      log.error("Something went wrong when terminating captain!");
    }
  }
};

export const loadFileFromFullPath = async (
  filePath: string,
): Promise<string> => {
  return fs.readFileSync(filePath, { encoding: "utf-8" }).toString();
};

export const saveFileToFullPath = async (
  filePath: string,
  content: string,
): Promise<Result<void>> => {
  try {
    fs.writeFileSync(filePath, content);
    return Ok(undefined);
  } catch (error) {
    return Err(error as Error);
  }
};
