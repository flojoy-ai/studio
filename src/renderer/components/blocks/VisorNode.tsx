import { CustomNodeProps } from "@/renderer/types/node";
import NodeWrapper from "@/renderer/components/common/NodeWrapper";
import clsx from "clsx";
import HandleComponent from "@/renderer/components/common/HandleComponent";
import Scatter from "@/renderer/assets/blocks/Scatter";
import { memo, useEffect, useMemo, useState } from "react";
import { makePlotlyData } from "@/renderer/components/plotly/formatPlotlyData";
import PlotlyComponent from "@/renderer/components/plotly/PlotlyComponent";
import CompositePlot from "@/renderer/assets/blocks/CompositePlot";
import ProphetComponents from "@/renderer/assets/blocks/ProphetComponents";
import ProphetPlot from "@/renderer/assets/blocks/ProphetPlot";
import ArrayView from "@/renderer/assets/blocks/ArrayView";
import MatrixView from "@/renderer/assets/blocks/MatrixView";
import BigNumber from "@/renderer/assets/blocks/BigNumber";
import BoxPlot from "@/renderer/assets/blocks/BoxPlot";
import Histogram from "@/renderer/assets/blocks/Histogram";
import LineChart from "@/renderer/assets/blocks/LineChart";
import Scatter3D from "@/renderer/assets/blocks/3DScatter";
import Surface3D from "@/renderer/assets/blocks/3DSurface";
import Bar from "@/renderer/assets/blocks/Bar";
import Table from "@/renderer/assets/blocks/Table";
import Image from "@/renderer/assets/blocks/Image";
import MarkDownText from "@/renderer/components/common/MarkDownText";
import { useTheme } from "@/renderer/providers/themeProvider";
import PeakFinder from "@/renderer/assets/blocks/PeakFinder";
import RegionInspector from "@/renderer/assets/blocks/RegionInspector";
import TextView from "@/renderer/assets/blocks/TextView";
import Heatmap from "@/renderer/assets/blocks/Heatmap";
import { useNodeStatus } from "@/renderer/hooks/useNodeStatus";
import { NodeResizer, useUpdateNodeInternals } from "reactflow";
import { useAtomValue } from "jotai";
import { projectAtom } from "@/renderer/hooks/useFlowChartState";

const chartElemMap = {
  SCATTER: Scatter,
  HISTOGRAM: Histogram,
  LINE: LineChart,
  SURFACE3D: Surface3D,
  SCATTER3D: Scatter3D,
  BAR: Bar,
  TABLE: Table,
  IMAGE: Image,
  BOX: BoxPlot,
  BIG_NUMBER: BigNumber,
  MATRIX_VIEW: MatrixView,
  ARRAY_VIEW: ArrayView,
  PROPHET_PLOT: ProphetPlot,
  PROPHET_COMPONENTS: ProphetComponents,
  COMPOSITE: CompositePlot,
  TEXT_VIEW: TextView,
  EXTREMA_DETERMINATION: PeakFinder,
  REGION_PROPERTIES: RegionInspector,
  HEATMAP: Heatmap,
};

const VisorNode = ({ selected, id, data }: CustomNodeProps) => {
  const { resolvedTheme } = useTheme();
  const { nodeRunning, nodeError, nodeResult } = useNodeStatus(data.id);

  const plotlyFig = nodeResult?.result?.plotly_fig;
  const textBlob = nodeResult?.result?.text_blob;

  const plotlyData = useMemo(
    () =>
      plotlyFig ? makePlotlyData(plotlyFig.data, resolvedTheme, true) : null,
    [plotlyFig, resolvedTheme],
  );

  const project = useAtomValue(projectAtom);
  const [dimensions, setDimensions] = useState({ width: 225, height: 225 });

  const updateNodeInternals = useUpdateNodeInternals();

  useEffect(() => {
    // Weird hack to make it properly set the dimensions when loading an app...
    // I tried like 10 different things but this is the only thing that works without being crazy slow
    const node = project.rfInstance?.nodes.find((n) => n.id === id);
    if (node?.width && node?.height) {
      setDimensions({ width: node.width, height: node.height });
    }
  }, [id, project]);

  const ChartIcon = chartElemMap[data.func];
  const iconSideLength = Math.min(dimensions.width, dimensions.height);

  return (
    <>
      <NodeResizer
        minWidth={225}
        minHeight={225}
        isVisible={selected}
        lineClassName="p-1"
        handleClassName="p-1"
        onResizeEnd={(e, params) => {
          setDimensions({ width: params.width, height: params.height });
          updateNodeInternals(id);
        }}
      />
      <NodeWrapper
        nodeError={nodeError}
        data={data}
        selected={selected}
        style={dimensions}
      >
        <div
          className={clsx(
            "rounded-2xl bg-transparent",
            { "shadow-around shadow-accent2": nodeRunning || selected },
            { "shadow-around shadow-red-700": nodeError },
          )}
          style={dimensions}
        >
          {plotlyData && (
            <PlotlyComponent
              data={plotlyData}
              id={data.id}
              layout={plotlyFig?.layout ?? {}}
              useResizeHandler
              isThumbnail
              style={dimensions}
            />
          )}
          {textBlob && <MarkDownText text={textBlob} isThumbnail />}
          {!plotlyData && !textBlob && (
            <>
              {ChartIcon ? (
                <ChartIcon width={iconSideLength} height={iconSideLength} />
              ) : (
                <div className="flex h-full w-full items-center justify-center break-all rounded-lg border-2 border-accent2 bg-accent2/5 p-2 text-center text-2xl font-bold tracking-wider text-accent2">
                  {data.label}
                </div>
              )}
            </>
          )}
          <HandleComponent data={data} variant="accent2" />
        </div>
      </NodeWrapper>
    </>
  );
};

export default memo(VisorNode);
