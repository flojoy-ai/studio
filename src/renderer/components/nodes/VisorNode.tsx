import { CustomNodeProps, ElementsData } from "@src/types/node";
import NodeWrapper from "@src/components/common/NodeWrapper";
import clsx from "clsx";
import HandleComponent from "@src/components/common/HandleComponent";
import Scatter from "@src/assets/nodes/Scatter";
import { memo, useEffect, useMemo, useState } from "react";
import { makePlotlyData } from "@src/components/plotly/formatPlotlyData";
import PlotlyComponent from "@src/components/plotly/PlotlyComponent";
import CompositePlot from "@src/assets/nodes/CompositePlot";
import ProphetComponents from "@src/assets/nodes/ProphetComponents";
import ProphetPlot from "@src/assets/nodes/ProphetPlot";
import ArrayView from "@src/assets/nodes/ArrayView";
import MatrixView from "@src/assets/nodes/MatrixView";
import BigNumber from "@src/assets/nodes/BigNumber";
import BoxPlot from "@src/assets/nodes/BoxPlot";
import Histogram from "@src/assets/nodes/Histogram";
import LineChart from "@src/assets/nodes/LineChart";
import Scatter3D from "@src/assets/nodes/3DScatter";
import Surface3D from "@src/assets/nodes/3DSurface";
import Bar from "@src/assets/nodes/Bar";
import Table from "@src/assets/nodes/Table";
import Image from "@src/assets/nodes/Image";
import MarkDownText from "@src/components/common/MarkDownText";
import { useTheme } from "@src/providers/themeProvider";
import PeakFinder from "@src/assets/nodes/PeakFinder";
import RegionInspector from "@src/assets/nodes/RegionInspector";
import TextView from "@src/assets/nodes/TextView";
import Heatmap from "@src/assets/nodes/Heatmap";
import { useNodeStatus } from "@src/hooks/useNodeStatus";
import { NodeResizer, useReactFlow } from "reactflow";

const chartElemMap = {
  SCATTER: Scatter,
  HISTOGRAM: Histogram,
  LINE: LineChart,
  SURFACE3D: Surface3D,
  SCATTER3D: Scatter3D,
  BAR: Bar,
  TABLE: Table,
  IMAGE: Image,
  BOX: BoxPlot,
  BIG_NUMBER: BigNumber,
  MATRIX_VIEW: MatrixView,
  ARRAY_VIEW: ArrayView,
  PROPHET_PLOT: ProphetPlot,
  PROPHET_COMPONENTS: ProphetComponents,
  COMPOSITE: CompositePlot,
  TEXT_VIEW: TextView,
  EXTREMA_DETERMINATION: PeakFinder,
  REGION_PROPERTIES: RegionInspector,
  HEATMAP: Heatmap,
};

const PLOT_ASPECT_RATIO = 1.4;

const VisorNode = ({ selected, data, id }: CustomNodeProps) => {
  const { resolvedTheme } = useTheme();
  const { nodeRunning, nodeError, nodeResult } = useNodeStatus(data.id);
  const { setNodes } = useReactFlow<ElementsData>();

  const plotlyFig = nodeResult?.result?.plotly_fig;
  const textBlob = nodeResult?.result?.text_blob;

  const plotlyData = useMemo(
    () =>
      plotlyFig ? makePlotlyData(plotlyFig.data, resolvedTheme, true) : null,
    [plotlyFig, resolvedTheme],
  );

  const [dimensions, setDimensions] = useState({ width: 225, height: 225 });

  const ChartIcon = chartElemMap[data.func];

  useEffect(() => {
    setNodes((nodes) => nodes.map(
      n => {
        if (n.id === id) {
          return {
            ...n,
            dimensions: {
              width: dimensions.width * PLOT_ASPECT_RATIO,
              height: dimensions.height
            }
          }
        }
        return n;
      }
    ))
  }, [plotlyData])

  return (
    <>
      <NodeResizer
        minWidth={225}
        minHeight={225}
        keepAspectRatio
        onResize={(_, params) => {
          const newSize = Math.max(params.width, params.height);
          setDimensions({ width: params.width, height: newSize });
        }}
        isVisible={selected}
      />
      <NodeWrapper nodeError={nodeError} data={data} selected={selected} style={dimensions}>
        <div
          className={clsx(
            "rounded-2xl bg-transparent",
            { "shadow-around shadow-accent2": nodeRunning || selected },
            { "shadow-around shadow-red-700": nodeError },
          )}
        >
          {plotlyData && (
            <PlotlyComponent
              data={plotlyData}
              id={data.id}
              layout={plotlyFig?.layout ?? {}}
              useResizeHandler
              isThumbnail
              style={{
                width: dimensions.width * PLOT_ASPECT_RATIO,
                height: dimensions.height
              }}
            />
          )}
          {textBlob && <MarkDownText text={textBlob} isThumbnail />}
          {!plotlyData && !textBlob && (
            <>
              {ChartIcon ? <ChartIcon className="w-full h-full" /> : (
                <div className="flex items-center justify-center break-all rounded-lg border-2 border-accent2 bg-accent2/5 p-2 text-center text-2xl font-bold tracking-wider text-accent2 w-full h-full">
                  {data.label}
                </div>
              )}
            </>
          )}
          <HandleComponent data={data} variant="accent2" />
        </div>
      </NodeWrapper>
    </>
  );
};

export default memo(VisorNode);
