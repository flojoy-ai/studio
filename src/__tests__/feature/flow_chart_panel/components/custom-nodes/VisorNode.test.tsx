import { CustomNodeProps } from "@src/feature/flow_chart_panel/types/CustomNodeProps";
import VisorNode from "@src/feature/flow_chart_panel/components/custom-nodes/VisorNode";
import { renderWithTheme } from "@src/__tests__/__utils__/utils";

const handleRemove = jest.fn();

const props: CustomNodeProps = {
  data: {
    id: "test-id",
    label: "test",
    func: "test",
    type: "test",
    ctrls: {},
    path: "",
    inputs: [
      {
        id: "test",
        name: "test",
        type: "test",
        desc: null,
      },
    ],
  },
  handleRemove,
};

jest.mock("@hooks/useFlowChartState");

jest.mock("@feature/flow_chart_panel/components/HandleComponent", () => {
  const mockChildren = jest
    .fn()
    .mockReturnValue(<div data-testid="handle-component" />);
  return { __esModule: true, default: mockChildren };
});

jest.mock("@src/hooks/useSocket");

describe("VisorNode", () => {
  it("checks the snapshot", () => {
    const { container } = renderWithTheme(<VisorNode {...props} />);
    expect(container).toMatchSnapshot();
  });
  it.each([
    ["SCATTER", "scatter-svg"],
    ["HISTOGRAM", "histogram-svg"],
    ["LINE", "line-svg"],
    ["SURFACE3D", "surface3D-svg"],
    ["SCATTER3D", "scatter3d-svg"],
    ["BAR", "barchart-svg"],
  ])("should contain component: %p", (functionName, testId) => {
    const { getByTestId } = renderWithTheme(
      <VisorNode
        data={{ ...props.data, func: functionName }}
        handleRemove={handleRemove}
      />
    );

    const component = getByTestId(testId);

    expect(component).toBeInTheDocument();
  });
});
