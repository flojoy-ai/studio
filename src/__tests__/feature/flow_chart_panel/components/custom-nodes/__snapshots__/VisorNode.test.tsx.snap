// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VisorNode checks the snapshot 1`] = `
<div>
  <div
    class="mantine-n8y7xk"
    data-testid="node-wrapper"
  >
    <div
      class="mantine-1avyp1d"
    >
      <div
        class="mantine-yenosr"
        style="padding: 0px 0px 8px 0px;"
      >
        <div
          class="mantine-1b6b4my"
        >
          <div
            data-testid="handle-component"
          />
        </div>
      </div>
    </div>
  </div>
  ,
</div>
`;
